### PR TITLE
feat(slack): allow join from private post link

### PR DESCRIPTION
**Pull request by capJavert**
- forward squad admin token as `?jt=token&type=squad&source=handle` which frontend can parse and then redirect to proper squad invite page
- also forwards all utm_ params
- commit for frontend implementation of join and redirect https://github.com/dailydotdev/apps/pull/3315/commits/1a616aee258922c2f54163e68bd34eebe1ed3d59
  - reuses referral logic (attributed to squad admin)
  - after squad join redirects back to post from which user joined
  - also preserves all utm_ params

### CODE REVIEW

File: src/workers/postAddedSlackChannelSend.ts [hunk 1]
```diff
@@ -38,6 +39,24 @@ export const postAddedSlackChannelSendWorker: TypedWorker<'api.v1.post-visible'>
         ]);
         const source = await post.source;
 
+        let adminJoinToken: string | undefined;
```
Comment (sshanzel) [inline src/workers/postAddedSlackChannelSend.ts:42]: Is it not automatically inferred to be `| undefined` if we don't define a value for it?

File: src/workers/postAddedSlackChannelSend.ts [hunk 2]
```diff
@@ -38,6 +39,24 @@ export const postAddedSlackChannelSendWorker: TypedWorker<'api.v1.post-visible'>
...
+
+        if (source.type === SourceType.Squad) {
+          const admin = await con.getRepository(SourceMember).findOne({
+            where: {
```
Comment (sshanzel) [inline src/workers/postAddedSlackChannelSend.ts:46]: Do we have to add the select statement?

**Review by sshanzel — COMMENTED** [review]
Looks amazing. I think we can add a test to verify that the token is sent along with the other parameters.

File: src/workers/postAddedSlackChannelSend.ts (hunk 1 above)
Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts:42]: nope, it just says string

File: src/workers/postAddedSlackChannelSend.ts (hunk 2 above)
Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts:46]: for sure!

Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts]: Added in 30b47a9a41b25ba688cb2a5c58ac8d5914e674cf together with test

File: __tests__/workers/postAddedSlackChannelSend.ts [hunk 3]
```diff
@@ -118,6 +161,49 @@ describe('postAddedSlackChannelSend worker', () => {
...
+    });
+    expect(chatPostMessage).toHaveBeenCalledWith({
+      channel: '1',
+      attachments: [
+        {
+          author_icon: 'https://app.daily.dev/apple-touch-icon.png',
+          author_name: 'daily.dev',
+          image_url: 'https://daily.dev/image.jpg',
+          title: 'Squad Channel Post 1',
+          title_link:
```
Comment (rebelchris) [inline __tests__/workers/postAddedSlackChannelSend.ts:198]: Do you maybe also want to test public squads for the diff or we already do so?

File: src/workers/postAddedSlackChannelSend.ts [hunk 4]
```diff
@@ -38,6 +39,27 @@ export const postAddedSlackChannelSendWorker: TypedWorker<'api.v1.post-visible'>
         ]);
         const source = await post.source;
 
+        let adminJoinToken: string | undefined;
+
+        if (source.type === SourceType.Squad) {
```
Comment (rebelchris) [inline src/workers/postAddedSlackChannelSend.ts:44]: Isn't it only needed for private squads?

**Review by rebelchris — APPROVED** [review]
More of a question

File: src/workers/postAddedSlackChannelSend.ts (hunk 4 above)
Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts:44]: Yeah good catch, for now actually better to specify the private check as well. I'll do that.

File: __tests__/workers/postAddedSlackChannelSend.ts (hunk 3 above)
Comment (capJavert) [reply __tests__/workers/postAddedSlackChannelSend.ts:198]: When I add the the private check in the worker the response handling will be the same as for source so no need I think. https://github.com/dailydotdev/daily-api/pull/2113#discussion_r1709237010

**Review by sshanzel — APPROVED** [review]
Nice. Looks amazing 🚀

File: src/workers/postAddedSlackChannelSend.ts (hunk 4 above)
Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts]: Solved in bd11f05533a7bd445c602bd8c69e12344ea8d816

File: __tests__/workers/postAddedSlackChannelSend.ts (hunk 3 above)
Comment (capJavert) [reply __tests__/workers/postAddedSlackChannelSend.ts:198]: Solved in bd11f05533a7bd445c602bd8c69e12344ea8d816

**Comment by capJavert** [comment]
@idoshamun if I can get your sanity check as well!

File: src/workers/postAddedSlackChannelSend.ts [hunk 5]
```diff
@@ -63,8 +85,16 @@ export const postAddedSlackChannelSendWorker: TypedWorker<'api.v1.post-visible'>
...
+
+              if (adminJoinToken) {
+                postLinkUrl.searchParams.set('jt', adminJoinToken);
+                postLinkUrl.searchParams.set('source', source.handle);
+                postLinkUrl.searchParams.set('type', source.type);
+              }
+
               const postLink = addNotificationUtm(
-                postLinkPlain,
+                postLinkUrl.toString(),
```
Comment (idoshamun) [inline src/workers/postAddedSlackChannelSend.ts:97]: Let's shorten the link to obscure this info and make it clean?

Comment (capJavert) [reply src/workers/postAddedSlackChannelSend.ts:97]: Was thinking about it, it's already obscured in the message markdown itself?

image

if we don't have to add another request to shortening service maybe better, wdyt?

**Pull request merged**